### PR TITLE
BUGFIX: Reduce the number of NodeTypes to run policy check against #2784

### LINKED ISSUE #2784: Improve performance of /neos/ui-services/get-additional-node-metadata endpoint
### Description

The endpoint `/neos/ui-services/get-additional-node-metadata` is called by the backend multiple times and can take many seconds to finish. In our case we have around 600 NodeTypes and this request can take 1-2 minutes.

The input for the call is a set of nodes which are processed here https://github.com/neos/neos-ui/blob/master/Classes/Controller/BackendServiceController.php#L465.

The runtime of this method depends on
* the number of nodes given as the input
* the number of NodeTypes configured
* the time needed for policy checks

Most of the time is consumed by https://github.com/neos/neos-ui/blob/master/Classes/Service/NodePolicyService.php#L109.

I'll provide a PR to reduce the number of NodeTypes processed during the policy checking.

### Steps to Reproduce

Have a look at the debug console in the neos backend of your browser. In bigger projects this request often takes a few seconds.

### Affected Versions

Neos: 4.x and up

UI: 4.x and up

### PR DESCRIPTION
**Pull request by gjwnc**
fixes #2784

**What I did**

In `getDisallowedNodeTypes(NodeInterface $node)` the list of configured NodeTypes is filtered using the NodeType constraints of the given `$node`. Abstract NodeTypes are still included which might be relevant for the policy check.

**How I did it**

Before running the policy check, `array_filter` is called used on all NodeTypes (`$this->nodeTypeManager->getNodeTypes()`). The filter method then checks the constraints depending if `$node` is auto-created or if the NodeType to filter is a superType or the NodeType of the node itself.

**How to verify it**

One need to compare the runtime of of the /get-additional-node-metadata endpoint in the Neos Backend UI on a project with configured CreateNodePrivilege policy.

This is a screenshot for a Neos 4.3 project without the code changes: 
image

And this one has the code changes in place:
image

### CODE REVIEW

File: Classes/Service/NodePolicyService.php [hunk 1]
```diff
@@ -114,14 +115,68 @@ public function getDisallowedNodeTypes(NodeInterface $node): array
...
+        $parentNodeType = null;
+        try {
+            $parentNode = $node->findParentNode();
+            $parentNodeType = $parentNode->getNodeType();
+        } catch (NodeException $exception) {
+            // no parentNode found
+        }
+
+        // check if the node is auto-created
+        $isAutCreated   = false;
```
Comment (dimaip) [inline Classes/Service/NodePolicyService.php:139]: isAutoCreated?

File: Classes/Service/NodePolicyService.php [hunk 2]
```diff
@@ -114,14 +115,68 @@ public function getDisallowedNodeTypes(NodeInterface $node): array
...
+            // check if the nodeType is the same as that of the current $node's nodeType
+            elseif ($nodeType->getName() == $nodeNodeType->getName()) {
+                return true;
+            }
+
+            // ignore other nodeType
+            return false;
+        };
+
+        $nodeTypes = array_filter($this->nodeTypeManager->getNodeTypes(), $constraintAndSuperTypeFilter);
```
Comment (dimaip) [inline Classes/Service/NodePolicyService.php:169]: For my taste I'd call it smth like `filteredNodeTypes`

**Comment by dimaip** [comment]
Can't really say something meaningful about this, I guess only @kitsunet can...

File: Classes/Service/NodePolicyService.php (hunk 1 above)
Comment (gjwnc) [reply Classes/Service/NodePolicyService.php]: Thanks! I've fixed it

File: Classes/Service/NodePolicyService.php (hunk 2 above)
Comment (gjwnc) [reply Classes/Service/NodePolicyService.php]: Yes, you are right. I've changed the variable name according to your suggestion.

**Comment by markusguenther** [comment]
@gjwnc Hi,

I tried to update the PR to the latest 5.3, but seems that I as team member has no permission to change the PR :/
Are you willing to rebase to origin/5.3 or should I take your commits and open a new PR?

Thanks for the effort :)

Screenshot 2021-10-15 at 13 39 57

**Comment by markusguenther** [comment]
Thanks for updating the branch 💚

**Comment by Sebobo** [comment]
Thx, will try the patch in a larger project and test it.

File: Classes/Service/NodePolicyService.php [hunk 3]
```diff
@@ -114,14 +115,68 @@ public function getDisallowedNodeTypes(NodeInterface $node): array
...
+        $generateSuperTypes = function (array $nodeTypes, &$superTypes) use (&$generateSuperTypes) {
+            foreach ($nodeTypes as $nodeType) {
+                $superTypes[$nodeType->getName()] = $nodeType;
+                $generateSuperTypes($nodeType->getDeclaredSuperTypes(), $superTypes);
+            }
+        };
+        $generateSuperTypes($nodeNodeType->getDeclaredSuperTypes(), $superTypes);
+
+        $parentNodeType = null;
+        try {
```
Comment (Sebobo) [inline Classes/Service/NodePolicyService.php:131]: Couldn't this try/catch be replaced by a `getParent` and checking whether its not null?

Comment (gjwnc) [reply Classes/Service/NodePolicyService.php:131]: Yes that would be possible, but `getParent()` is deprecated and in the comment for `getParent()` it is mentioned that you should use `findParentNode()`.

Comment (Sebobo) [reply Classes/Service/NodePolicyService.php:131]: True, just wondered if the try/catch stuff might have a negative impact again. But not sure how many times the method is actually called.

**Comment by Sebobo** [comment]
Did you project use a lot of node related policies? I could see a small difference of ~15% in my customers project which is quite large, but they don't use node policies.

**Comment by gjwnc** [comment]
We use EditNodePropertyPrivilege and Create|RemoveNodePrivilege very often in the project because we limit access to the site and language dimension.

What do you mean by difference? The returned array of the `getDisallowedNodeTypes()`?  
It's quite some time, that I worked on this, but as I read it now again, I think in the final result, I miss the set of nodeTypes the policy check is not running against. Because this set is disallowed by definition for the given node's nodeType, right?

**Comment by Sebobo** [comment]
I meant the performance improvement is ~15%, but I have to make more checks with pages that have more nodes.

And regarding the set of NodeTypes. I think you also include them in line 169.

**Comment by mhsdesign** [comment]
should/can this new code be extracted into a new method? just a lot going on...

**Comment by gjwnc** [comment]
@mhsdesign Do you mean split `getDisallowedNodeTypes()` into a protected methods or creating a completely separate new method along the original `getDisallowedNodeTypes()`?  
I would just move the anonymous functions for `$generateSuperTypes` and `$constraintAndSuperTypeFilter` to separate methods.
Since this is more than a year old I'm questioning if it will be merge sometime soon - if it's worth it currently. It's basically a nice to have I noticed quite some time ago.

File: Classes/Service/NodePolicyService.php (hunk 3 above)
Comment (mhsdesign) [reply Classes/Service/NodePolicyService.php:131]: https://php.watch/articles/php-exception-performance

just less performant, when an exception will be thrown.

Comment (mhsdesign) [reply Classes/Service/NodePolicyService.php:131]: but an exception will only be thrown when `$this->isRoot() === true`:
https://github.com/neos/neos-development-collection/blob/c14cc4ab46e7d9be75157bb5aa80f96fcc2dff14/Neos.ContentRepository/Classes/Domain/Model/Node.php#L2026

so we could instead check for that first and never get an exception.

**Review by Sebobo — APPROVED** [review]

**Comment by Sebobo** [comment]
I'm fine with merging this change, but it would be great if someone could test it in a project where the policies are extensively used.

**Review by kitsunet — APPROVED** [review]
Great idea, I needed a couple of reads, but seems sound to me and works. So I would agree to merge but have to bring up the question of version. I am not sure this can be declared a bug, strictly speaking. Performance is a feature, and such optimizations always have the risk of breaking something :D Would at least want to hear some other opinions, but would probably otherwise merge in ~24h as I feel it's a good change.

**Comment by Sebobo** [comment]
I would feel better to merge into 8.0 honestly.

**Comment by mhsdesign** [comment]
So better merge this into 8.0 as Feature or as (Performance) Bug in 5.3?

**Pull request merged**